scripts/mysql_system_tables.sql

```diff
@@ -86,6 +96,8 @@ CREATE TABLE IF NOT EXISTS proc (db char(64) collate utf8_bin DEFAULT '' NOT NUL
 
 CREATE TABLE IF NOT EXISTS procs_priv ( Host char(60) binary DEFAULT '' NOT NULL, Db char(64) binary DEFAULT '' NOT NULL, User char(80) binary DEFAULT '' NOT NULL, Routine_name char(64) COLLATE utf8_general_ci DEFAULT '' NOT NULL, Routine_type enum('FUNCTION','PROCEDURE','PACKAGE','PACKAGE BODY') NOT NULL, Grantor char(141) DEFAULT '' NOT NULL, Proc_priv set('Execute','Alter Routine','Grant') COLLATE utf8_general_ci DEFAULT '' NOT NULL, Timestamp timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP, PRIMARY KEY (Host,Db,User,Routine_name,Routine_type), KEY Grantor (Grantor) ) engine=MyISAM CHARACTER SET utf8 COLLATE utf8_bin   comment='Procedure privileges';
 
+CREATE TABLE IF NOT EXISTS reverse_procs_priv ( Host char(60) binary DEFAULT '' NOT NULL, Db char(64) binary DEFAULT '' NOT NULL, User char(80) binary DEFAULT '' NOT NULL, Routine_name char(64) COLLATE utf8_general_ci DEFAULT '' NOT NULL, Routine_type enum('FUNCTION','PROCEDURE') NOT NULL, Grantor char(141) DEFAULT '' NOT NULL, Proc_priv set('Execute','Alter Routine','Grant') COLLATE utf8_general_ci DEFAULT '' NOT NULL, Timestamp timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP, PRIMARY KEY (Host,Db,User,Routine_name,Routine_type), KEY Grantor (Grantor) ) engine=MyISAM CHARACTER SET utf8 COLLATE utf8_bin   comment='Procedure privileges';
+
```

Comment:
Routine_type doesn't match procs_priv.

Would adding a ```grant ENUM('grant', 'deny') DEFAULT 'grant'```column to existing tables be easier to maintiain and less lookups during authentication?

Reply:
"* I'm not a great fan of duplicating all tables, I'd rather reuse existing ones"

@vuvova already suggested this https://lists.launchpad.net/maria-developers/msg11286.html

Reply:
I've modified the routine_type enum as per the comment and have added (MDEV-14443) to the commit message. The idea behind creating duplicate tables was to be able to make 'deny' dominate over grant (i.e., if a particular privilege has a 'Y' in the grants table and 'Y' in the deny table, then deny takes precedence). Keeping both grant and deny in the same table wouldn't allow us to do that as the primary key also would be the same for two conflicting entries (can't have 'y' for grants and 'y' for deny for a record in the same table). We want to test the behaviour of reverse privileges by making all_load read these tables on calling flush privileges

Reply:
nit a) There's never brackets around MDEV, follow the convention used, ```MDEV-XXXX: ....``` or just ```MDEV .....```.

b) use ```git rebase -i {base of branch}```, to squash up fixes. 9 commits with similar/same subjects in hard to review. When you are happy how it looks in ```git log```, then you can ```git push --force``` them to the PR.

c) I see further in the thread that @cvicentiu has lead you down this path, I think its a mistake and that precluding identical grants/revoke at a sql level (appropriate unique key) has benefits in maintenance and simplicty. However but I'm not your mentor, haven't read your proposal so I'll leave this go. :-)

d) for a duplicate table approach, how does using `create table if not exists reverse_db like db` sound?

Reply:
Thanks a lot for the review, I have altered the commit message and create table statements accordingly. I will squash up the commits to a single one as well.

Reply:
@grooverdan: Regarding point c)

I am not satisfied with the doubling of privilege tables either. We discussed both approaches. The main takeaway is that the code that handles this part is rather trivial to change, regardless how we chose to implement it. I wanted to make a decision so we can move forward for now. Later we can chose to revise this and implement it in a more user friendly fashion.

Here's the main use case that I didn't like with using just mysql.user table:
To store both positive and negative grants for a user, you would need two entries into the same table, most likely differentiated by a different column, let's name it *is_deny*. If somebody does select user, host from mysql.user, just to check which users are in the system, this might break those scripts. I didn't want to potentially change behaviour of mysql.user table that has remained stable since basically forever.

Do you @grooverdan find this problematic use case a non-problem? Am I being too paranoid? Any sort of suggestions are entirely welcome as we're treading into unknown territory. :) I've gotten the feedback that doubling the tables is not something people are a "fan of" but I didn't quite get the exact reasons for that. Being more precise would help narrow down the actual problems and maybe lead us to a better solution.

Thanks!

Reply:
Is the case for denials in mysql.user things like Super but deny alter routine? If so I'd look at expanding the enums.

For doubling the tables I was concerned this would double lookups for checking access. A negative ACL is increasing lookups anyway I'm assuming.

I was thinking looking up user,host, and some sort of privileged would result in one query would result in one row (indexed) per table and it would be a deny or not. Am I wrong here? Maybe this gets cached well, acl cache isn't something I've looked at either.